README.md
# luminave-thorium

Create a connection between [luminave](https://github.com/NERDDISCO/luminave) and the [Thorium Simulator](https://thoriumsim.com/) using GraphQL.

[![Build Status](https://travis-ci.org/NERDDISCO/luminave-thorium.svg?branch=master)](https://travis-ci.org/NERDDISCO/luminave-thorium)

## Table of Contents

<!-- toc -->

- [Setup](#setup)
- [Config](#config)
  * [Example .env](#example-env)
- [Use with luminave & Thorium](#use-with-luminave--thorium)
  * [General Workflow](#general-workflow)
  * [GraphQL](#graphql)
- [Troubleshooting](#troubleshooting)
  * [No data received from Thorium](#no-data-received-from-thorium)
  * [One action in Thorium is triggering more than 1 changes in luminave e.g. the same scene multiple times](#one-action-in-thorium-is-triggering-more-than-1-changes-in-luminave-eg-the-same-scene-multiple-times)
- [API Documentation](#api-documentation)
- [Data provided by Thorium](#data-provided-by-thorium)

<!-- tocstop -->

---

## Setup

* Clone [this repository](https://github.com/NERDDISCO/luminave-thorium)
* Install the dev dependencies by executing `npm install` inside the *luminave-thorium* folder

---

## Config

If you want to change the `protocol`, `host` & `port` of *luminave-server* or *thorium*, you have to create a `.env` file inside the *luminave-thorium* folder. 

### Example .env

```bash
# Protocol of the luminave-server, default is HTTP
#PROTOCOL_LUMINAVE_SERVER=http

# Host of the luminave-server
HOST_LUMINAVE_SERVER=localhost

# Port of the luminave-server
PORT_LUMINAVE_SERVER=4000

# Protocol of Thorium, default is HTTP
#PROTOCOL_THORIUM=https
# Host of Thorium
# Leave it undefined if you want to have it auto detected with bonjour
#HOST_THORIUM=undefined
# Port of Thorium
# leave it undefined if you want to have it auto detected with bonjour
#PORT_THORIUM=undefined

# ID of the Thorium client that is used by the Thorium Server to
# identify this client as the one that can control the lighting
# Note: As of now this will only work if the letters 'ECS' are part of the clientId
# Note: Only change if you know what you are doing
#THORIUM_CLIENTID='ECS'

# ID that will be used in luminave to identify the animation that will be used to 
# dynamically update it's keyframes
# Note: Only change if you know what you are doing
#LUMINAVE_ANIMATIONID=luminave-thorium-dynamic-animation-1337
```

---

## Use with luminave & Thorium

* Start with `npm start`
* Start [Thorium server](https://github.com/Thorium-Sim/thorium) with `npm start`
* Create a flight in Thorium at http://localhost:3000
* Add the luminave-thorium as a client with the name `ECS-luminave-thorium` to the created flight

That's it, luminave-thorium is now ready to receive data from Thorium to send it over to [luminave-server](https://github.com/NERDDISCO/luminave-server). 


### General Workflow

* The app tries to find the Thorium Server running in the local network
* When it finds it it tries to setup a connection to it and registers itself as a client called "ECS"
* When this was successful, it subscribes to new messages on the simulator and the lighting changes (thorium-lighting)
* Those updates are send via a mutation to the luminave-server


### GraphQL

* Use the GraphiQL on http://localhost:3001/graphiql to test your queries

--- 

## Troubleshooting

### No data received from Thorium

* Check if the client IP changed, if that happened you have to restart Thorium & luminave-thorium so it registers itself with the new IP of your computer


### One action in Thorium is triggering more than 1 changes in luminave e.g. the same scene multiple times

* Restart thorium-client and refresh Thorium in the browser



---

## API Documentation

Can be found on [API documentation](docs/API.md).

--- 

## Data provided by Thorium

[Detailed explanation on how this data could be interpreted.](https://github.com/Thorium-Sim/thorium/issues/1645#issuecomment-445867388)

```graphql
query Simulators($simulatorId: String) {
  simulators(id: $simulatorId) {
    id
    lighting {
      intensity # [0 - 1] How bright the lights should be
      action # One of 'normal', 'fade', 'shake', 'strobe', 'oscillate'
      actionStrength # [0 - 1] How intense the action is. How quickly it shakes, or how fast it strobes or oscillates
      transitionDuration # How long a 'fade' action should last in milliseconds
      color # A suggested color for the lights
    }
  }
}
```


docs/CHANGELOG.md
# [1.5.0](https://github.com/NERDDISCO/luminave-thorium/compare/v1.4.0...v1.5.0) (2020-01-29)


### Features

* **config:** Change the protocol ([28fce05](https://github.com/NERDDISCO/luminave-thorium/commit/28fce05))

# [1.4.0](https://github.com/NERDDISCO/luminave-thorium/compare/v1.3.0...v1.4.0) (2019-10-16)


### Bug Fixes

* **thorium-lighting:** Only update scene when there was a change ([c66ad6e](https://github.com/NERDDISCO/luminave-thorium/commit/c66ad6e)), closes [#10](https://github.com/NERDDISCO/luminave-thorium/issues/10)


### Features

* **thorium-lighting:** Map alert levels to scenes ([e5311d1](https://github.com/NERDDISCO/luminave-thorium/commit/e5311d1)), closes [#9](https://github.com/NERDDISCO/luminave-thorium/issues/9)

# [1.3.0](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.7...v1.3.0) (2019-10-14)


### Features

* **luminave-client:** Allow black as scene ([fff7e77](https://github.com/NERDDISCO/luminave-thorium/commit/fff7e77)), closes [#13](https://github.com/NERDDISCO/luminave-thorium/issues/13)

## [1.2.7](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.6...v1.2.7) (2019-10-14)


### Bug Fixes

* **thorium-lighting:** Unsubscribe from simulatorUpdate subscription ([5133898](https://github.com/NERDDISCO/luminave-thorium/commit/5133898)), closes [#10](https://github.com/NERDDISCO/luminave-thorium/issues/10)

## [1.2.6](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.5...v1.2.6) (2019-08-25)


### Bug Fixes

* **graphql:** Use correct data for initial query to Thorium ([e922214](https://github.com/NERDDISCO/luminave-thorium/commit/e922214))

## [1.2.5](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.4...v1.2.5) (2019-07-20)


### Bug Fixes

* **subscription:** Only use changed data for luminave ([a528ee6](https://github.com/NERDDISCO/luminave-thorium/commit/a528ee6)), closes [#5](https://github.com/NERDDISCO/luminave-thorium/issues/5)

## [1.2.4](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.3...v1.2.4) (2019-07-18)


### Bug Fixes

* **bonjour:** Find just one thorium server ([54e14df](https://github.com/NERDDISCO/luminave-thorium/commit/54e14df)), closes [#6](https://github.com/NERDDISCO/luminave-thorium/issues/6)

## [1.2.3](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.2...v1.2.3) (2019-06-10)


### Bug Fixes

* **graphql:** Updates the lighting query to use the correct variable type. ([d03e243](https://github.com/NERDDISCO/luminave-thorium/commit/d03e243))

## [1.2.2](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.1...v1.2.2) (2019-06-10)


### Bug Fixes

* **promises:** Catch errors in promises when doing GraphQL request ([57b30ef](https://github.com/NERDDISCO/luminave-thorium/commit/57b30ef))

## [1.2.1](https://github.com/NERDDISCO/luminave-thorium/compare/v1.2.0...v1.2.1) (2019-06-07)


### Bug Fixes

* **config:** Changed name of client to ECS-luminave-thorium ([ab7dab0](https://github.com/NERDDISCO/luminave-thorium/commit/ab7dab0))

# [1.2.0](https://github.com/NERDDISCO/luminave-thorium/compare/v1.1.0...v1.2.0) (2019-05-27)


### Features

* **config:** Added dotenv ([1c0dfe4](https://github.com/NERDDISCO/luminave-thorium/commit/1c0dfe4)), closes [#2](https://github.com/NERDDISCO/luminave-thorium/issues/2)
* **config:** Externalized configuration & added debugMode ([29bef08](https://github.com/NERDDISCO/luminave-thorium/commit/29bef08)), closes [#2](https://github.com/NERDDISCO/luminave-thorium/issues/2)
* **graphql:** Set animation in luminave-server ([9b4d4a5](https://github.com/NERDDISCO/luminave-thorium/commit/9b4d4a5))

# [1.1.0](https://github.com/NERDDISCO/luminave-thorium/compare/v1.0.0...v1.1.0) (2019-02-07)


### Features

* **graphql:** Create connection to luminave-server ([687379a](https://github.com/NERDDISCO/luminave-thorium/commit/687379a))

# 1.0.0 (2018-12-13)


### Bug Fixes

* **jsdoc:** Use correct files to generate API documentation ([c81049c](https://github.com/NERDDISCO/luminave-thorium/commit/c81049c))


### Features

* **hello world:** Let's connect luminave with Thorium ([b08cfae](https://github.com/NERDDISCO/luminave-thorium/commit/b08cfae))


src/config.js
import dotenv from 'dotenv'

// Parse the .env file
dotenv.config()

const config = {
  env: process.env.NODE_ENV || 'development',

  // protocol of the luminave-server
  protocolLuminaveServer: process.env.PROTOCOL_LUMINAVE_SERVER || 'http',
  // Host of the luminave-server
  hostLuminaveServer: process.env.HOST_LUMINAVE_SERVER || 'localhost',
  // Port of the luminave-server
  portLuminaveServer: parseInt(process.env.PORT_LUMINAVE_SERVER) || 4000,

  // Protocol of Thorium
  protocolThorium: process.env.PROTOCOL_THORIUM || 'http',
  // Host of Thorium
  // Leave it undefined if you want to have it auto detected with bonjour
  hostThorium: process.env.HOST_THORIUM || undefined,
  // Port of Thorium
  // leave it undefined if you want to have it auto detected with bonjour
  portThorium: parseInt(process.env.PORT_THORIUM) || undefined,

  // ID of the Thorium client that is used by the Thorium Server to
  // identify this client as the one that can control the lighting
  // Note: As of now this will only work if the letters 'ECS' are part of the clientId
  thoriumClientId: process.env.THORIUM_CLIENTID || 'ECS-luminave-thorium',

  // ID that will be used in luminave to identify the animation that will be used to 
  // dynamically update it's keyframes
  luminaveAnimationId: process.env.LUMINAVE_ANIMATIONID || 'luminave-thorium-dynamic-animation-1337'
}


// When in debug mode there will be more log messages
config.debugMode = config.env === 'development'

export default config
